Enable query buffers (temp tables) to support LIMIT, OFFSET and ORDER BY

hmmr opened the pull request:
RTS-733, RTS-870, RTS-1006

Depends on https://github.com/basho/riak_ql/pull/139, https://github.com/basho/riak_pb/pull/208; r_t module in https://github.com/basho/riak_test/pull/1152.

This PR implements https://github.com/basho/rfc/pull/5:
1. Introduces query buffers, a node-local temporary storage for SELECT queries, which can now be of arbitrary size and span arbitrary range (thus lifting the restriction on max subqueries).
2. Enables sorting of SELECT query results with the new keyword ORDER BY.
3. Enables the LIMIT and OFFSET keywords.
4. Allows reuse of query buffers for subsequent matching queries, controlled by the newly introduced flag `#tsqueryreq.allow_qbuf_reuse` (off by default).

Commit:
support for LIMIT, OFFSET, ORDER BY (via query buffers)

Including ASC/DESC and NULLS FIRST/LAST modifiers for ORDER BY
elements, and the ONLY keyword to prevent reuse of query buffers.

Commit:
use #tsqueryreq.allow_qbuf_reuse instead of ONLY keyword

also,

* straighten up case clauses around riak_kv_ts_svc:decode_query;
* bundle Cover and AllowQBufReuse as a proplist into Options.

hmmr commented:
Reverting to WIP status as I and @andytill have discovered issues around lifting the MaxSubqueries limitation.

Commit:
remove disused debugging aids in kv_qry_worker state

Commit:
throttle index_fsm breeding in qry_worker

Commit:
qry_buffers: single-quote ~s in formatted os:cmd

Commit:
qry_worker: remove unused defines, historical comments

Commit:
qry_worker: estimate projected query size, cancel if too large

hmmr commented:
In 6eca18d, I am replacing the limit on max subqueries with a total query result size limit. It is calculated as `TotalQuerySize = AverageChunkSize * NSubqueries` for regular queries (those not backed by the query buffers). For query buffer-based queries, `TotalQuerySize` is the LIMIT value times average chunk size.

If the total query result size is projected to exceed a certain, configurable value (hardcoded to be 5 MiB), the query is aborted.

The sizes are computed after incorporation of each chunk, starting from chunk 2, in and around https://github.com/basho/riak_kv/pull/1479/files#diff-85d69bbf5fbaa852650866d125af628aR287.

Commit:
qry_buffers: more cleanly separate ldb ops

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+        false ->
+            {reply, {error, bad_qbuf_ref}, State0};
+        #qbuf{is_ready = true} ->
+            %% qbuf is already ready: don't wait for the bext
```
'bext' should be 'next'

gordonguthrie commented on include/riak_kv_ts.hrl:
```diff
@@ -52,8 +52,8 @@
           'WHERE'       = []    :: [riak_ql_ddl:filter()],
-          'ORDER BY'    = []    :: [riak_kv_qry_compiler:sorter()],
-          'LIMIT'       = []    :: [riak_kv_qry_compiler:limit()],
+          'ORDER BY'            :: undefined | [riak_kv_qry_compiler:sorter()],
```
These elements are empty lists because the values are mini-LISP programmes that are to be evaluated elsewhere

hmmr replied:
Can you clarify on that? I see `ORDER BY` as a list of fields by which sorting is to be done, annotated with asc/desc and/or nulls first/last modifiers (`LIMIT` is a simple integer).

gordonguthrie replied:
They will all be implemented as operations going forward:

``` [
     {order_by, [<<"col1">>, <<"col4">>]},
     {as,           [{<<"col1">>, <<"new fancy name">>}]},
     {limit,        [1000]}
     ],
```

In this world a `NoOp` is `[ ]`

gordonguthrie commented on include/riak_kv_ts.hrl:
```diff
@@ -83,6 +83,27 @@
+          %% prolly a mistake to put this here - should be in DDL
+          local_key,
+          %% since v2
+          group_by = ?GROUP_BY_DEFAULT :: [{identifier, binary()}] | [{FieldPos::integer(), FieldName::binary()}]
```
This should be 'GROUP BY' by convention we preserve SQL clauses upper case in the SQL records - the rational is that these fields are semantically the same as the original SQL clauses (or can be so reconciled) whilst containing syntactic differences pertaining to execution. It should also be insert up into the record with the rest of the SQL clauses.

hmmr replied:
I agree in general but I would rather leave the rename to @andytill. It will obviously require a sweeping change all over the place. The current PR's diff is already big enough.

gordonguthrie commented on src/riak_kv_app.erl:
```diff
@@ -213,7 +213,7 @@ start(_Type, _StartArgs) ->
 
             riak_core_capability:register({riak_kv, sql_select_version},
-                                          [v2,v1],
+                                          [v3,v2,v1],
```
**Note** implies upgrade/downgrade

gordonguthrie commented on src/riak_kv_qry.erl:
```diff
@@ -281,22 +281,62 @@ count_to_position(Col, [_ | Rest], Pos) ->
+                              riak_kv_qry_queue:put_on_queue(self(), SubQueries, DDL, QBufRef))
+                        end,
+                    case maybe_create_query_buffer(SQL, length(SubQueries), DDL, []) of
+                        {error, _Reason} ->
```
If it fails, it fails and should be logged and reported as an error instead of being left to go on and fail somewhere else

hmmr replied:
Rewrote to make it a query-cancelling error (26648b15).

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+        ]).
+
+-export([
+         batch_put/2,           %% emulate INSERT (new Chunk collected by worker for a query Ref)
```
There are two seperate API's here:

Used API:
• `data_row`
• `fetch_limit`
• `qbuf_ref`
• `batch_put`
• `delete_qbuf`
• `set_ready_waiting_process`
• `get_or_create_qbuf`

Developer API:
• `get_qbuf_orig_query`
• `get_qbuf_size`
• `get_qbufs`
• `is_qbuf_ready`
• `set_qbuf_expiry`
• `make_qref`

If you have a developer/testing API you have a couple of options - certainly have a different export statement - and consider decorating the function names with things like `make_qref_DEV`

hmmr replied:
I have removed all unused, but exported, internal API functions, in d8b39a1.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+%% @private
+init(Options) ->
+    RootPath =
+        filename:join(
```
@The 'design' of the data directory is a UX thing (we expose stuff to customers/CSEs) so lets get @angrycub 's view on this.

Also the fileroot should be passed in as an option - this means that a test suite can be written which writes to /tmp or a memory disk or whatever...

angrycub replied:
I do feel like there are usability benefits for a configurable path. A user might want to store this data on AWS instance storage. Configurable paths would prevent symlink hell for users that want to take advantage.  A riak.conf default of `${platform_data_dir}/query_buffers` would align nicely with AAE data path configurations (https://github.com/basho/riak_kv/blob/develop/priv/riak_kv.schema#L123-L128).   Might be good to get an additional CSE eyeball on this.

hmmr replied:
It is now configurable in riak_kv.schema (26648b15).

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+
+schedule_tick(Pid) ->
+    Pid ! tick,
+    timer:sleep(1000),
```
Macroise 1000 and setup at to the top the rest of your timeouts, watermarks etc...

hmmr replied:
Macroised (20000d0). As for parameters, they are now configurable in riak_kv.schema (0c15cd0).

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+
+-spec handle_call(term(), pid() | {pid(), term()}, #state{}) -> {reply, term(), #state{}}.
+%% @private
+handle_call(get_qbufs, _From, State) ->
```
this handle_call/do_call pattern is a waste of space...

hmmr replied:
It's a choice between putting all code in `handle_call`s, or none. Given the size of `do_get_or_create_qbuf` and others, the first is clearly not a good option.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+    {noreply, State}.
+
+
+-spec handle_info(term(), #state{}) -> {noreply, #state{}}.
```
I would do the tick in a `call` or `cast` not `info`

`info` is for messages that your process gets in some only slightly expected way from any other process (or some random problem messages)

hmmr replied:
Did it via `gen_server:cast` in 2ee4fb1.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+  when OrderBy /= undefined ->
+    Part0 = erlang:phash2(From),
+    Part1 = erlang:phash2(ColNames),
+    Part2 = erlang:phash2(lists:sort(Where)),
```
The WHERE clause is canonicalised in the lexer/parser - all the clauses should be canonical coming out of that...

hmmr replied:
Done in 783870a.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+maybe_ensure_qref(?SQL_SELECT{allow_qbuf_reuse = true} = SQL, QBufs) ->
+    ensure_qref(SQL, QBufs);
+maybe_ensure_qref(?SQL_SELECT{allow_qbuf_reuse = false}, _QBufs) ->
+    AlwaysUniqueRef = crypto:rand_bytes(8),
```
Don't use hash for unique values - use References - first class Erlang terms

http://erlang.org/doc/man/erlang.html#make_ref-0

hmmr replied:
I need it to be a binary, the same size as that generated by `make_qref` (if only for it to be of the same type from dialyzer's POV).

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+
+maybe_inform_waiting_process(true, Pid, QBufRef) when is_pid(Pid) ->
+    lager:debug("notifying waiting pid ~p", [Pid]),
+    Pid ! {qbuf_ready, QBufRef};
```
@lehoff recommends not passing in PIDs but using fns like this as a reference

```
%% Caputure Pid in a closuse
RefToMeFn = fun(Msg) -> 
        Pid ! {qbuf_read, Msg),
        ok
end,
setup_query_buffer(Blah, Bleh, Bloh, RefToMeFn)
...
```

And then in your module you do

```
ok = ReplyFn(MyMsg)
```

hmmr replied:
Okay (7c81ca9).

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+%%-------------------------------------------------------------------
```
**Note** this module needs a test - prolly a common test or mebbies EQC

gordonguthrie replied:
What is the GUI design for specifying the total system resource limits as per @matthewvon 's spec for this? We need an associated `clique` change

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+                               expire_msec = ExpireMsec,  %% qbuf-specific, possibly overriden
+                               last_accessed = LastAccessed}}) ->
+                  ExpiresOn = advance_timestamp(LastAccessed, ExpireMsec),
+                  if ExpiresOn < Now ->
```
use a case statement

hmmr replied:
Replaced with case in 2aec11b.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+    %% i.e., without WHERE or GROUP BY). We can match on {ok, SQL}
+    %% because any errors will have been detected when the compilation
+    %% chain was run on the query in riak_kv_qry_compiler.
+    {ok, CompiledSelect} = riak_kv_qry_compiler:compile_select_clause(OrigDDL, OrigSQL),
```
This looks like a code smell - we have already compiled the SQL and we are now recompiling it to re-extract bits...

hmmr replied:
Done in ec4b352.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+         %% eleveldb
+         local_key     = make_lk_from_orderby(OrderBy)}.
+
+make_qbuf_id(From, Select, OrderBy) ->
```
This query buffer ID is not unique because the timestamp isn't - `os:now()` is guaranteed to be unique but has a cost and will be deprecated in more modern versions of Erlang

hmmr replied:
Added a sleep of 1 msec to emulate `erlang:now/1` in 2ea39988.

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+
+
+%% leveldb instance parameters
+-define(LDB_WRITE_BUFFER_SIZE, 10*1024*1024).  %% 10 M should be enough for everybody
```
What is this? How does it relate to @matthewvon 's params?

hmmr replied:
This setting sets the max amount of data to keep in memory before writing to disk happens.

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+-spec new_table(binary(), string()) -> {ok, eleveldb:db_ref()} | {error, term()}.
+new_table(Table, Root) ->
+    Path = filename:join(Root, binary_to_list(Table)),
+    _ = os:cmd(fmt("mkdir -p '~s'", [Path])),
```
Don't use `os:cmd`

http://erlang.org/doc/man/file.html#make_dir-1

hmmr replied:
We now have our very own `riak_kv_ts_util:rm_rf/1` (916c59e). For the mkdir case, there is `filelib:ensure_dir/1`.

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,815 @@
+          app_helper:get_prop_or_env(data_root, Options, eleveldb),  %% reuse eleveldb own root_path
+          "query_buffers"),
+    %% don't bother recovering any leftover tables
+    os:cmd(
```
Don't use `os:cmd`

http://erlang.org/doc/man/file.html#del_dir-1

hmmr replied:
Now using `riak_kv_ts_util:rm_rf/1` and `filelib:ensure_dir/1` (916c59e).

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+            lager:info("new LdbRef ~p in ~p", [LdbRef, Path]),
+            {ok, LdbRef};
+        {error, _} = ErrorReason ->
+            _ = os:cmd(fmt("rm -rf '~s'", [Path])),
```
Don't use `os:cmd` ;-)

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+-spec delete_table(binary(), eleveldb:db_ref(), string()) -> ok.
+delete_table(Table, LdbRef, Root) ->
+    ok = eleveldb:close(LdbRef),
+    _ = os:cmd(
```
Don't use `os:cmd` ;-)

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+            {error, ldb_put_failed}
+    end.
+
+maybe_negate([], desc, nulls_first) ->
```
`mabye_negate` is what you are doing - why you are doing it is `make_sorted_keys`

hmmr replied:
Renamed in 78747953.

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,172 @@
+            Decoded =
+                [sext:decode(Row) || Row <- lists:reverse(Fetched)],
+            {ok, Decoded}
+        %% {error, Reason} ->
```
If it ain't happening, delete it...

hmmr replied:
It was my attempt, inexpressive as it is, to deal with the fact that `eleveldb:fold` does not return any `{error, Reason}`. Refactored a little in 9215fd2801.

gordonguthrie commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -43,35 +45,61 @@
 -export_type([
     combinator/0,
-    limit/0,
+    limit/0, offset/0,
```
go vertical or go horizontal

gordonguthrie commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -43,35 +45,61 @@
+        ?SQL_SELECT{is_executable = false} = Q1) ->
     Mod = riak_ql_ddl:make_module_name(T),
-    case maybe_compile_group_by(Mod, compile_select_clause(DDL, Q1), Q1) of
+    case compile_order_by(
```
time to break up this set of nested mebbies into a single compile function that branches on matching heads

gordonguthrie commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -43,35 +45,61 @@
+
+non_unique_identifiers(FF) ->
+    Occurses = [{F, occurs(F, FF)} || {F, _, _} <- FF],
+    [F || {F, N} <- Occurses, N > 1].
```
Hooray @hmmr tripped up on his English [does small jig in seat] - not sure 'occurses' is a word ;-)

Revenge for https://github.com/basho/riak_shell/commit/83d7ca28d132af1046964f738b17c3ae6c4ee9de

hmmr replied:
**Abstract**

English influences in modern Erlang are subtle and not well understood.

**Introduction**

`Occurses` may not be correct English but it _is_ correct programmese.

**Discussion**

There is logic and consistency in calling `Poses` a list of positions, each denoted by variable `Pos` elsewhere in your code.

**Conclusions**

I beg to disagree.

gordonguthrie replied:
I would be Occurances ;-)

gordonguthrie commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -518,7 +543,7 @@ find_quantum_field_index_in_key2([_|Tail], Index) ->
 
 %%
-hash_timestamp_to_quanta(QField, QSize, QUnit, QIndex, MaxSubQueries, Where1) ->
+hash_timestamp_to_quanta(QField, QSize, QUnit, QIndex, Where1) ->
```
Need to test this and see if there is a max-max subqueries limit we should put in...

hmmr replied:
The max subqueries check was reinstated in bfb275b5 (initially in 3fc2471a, it was removed); it is now done earlier in the call chain. In 0c15cd06, it is made configurable in riak_kv.schema.

gordonguthrie commented on src/riak_kv_qry_compiler.erl:
```diff
@@ -547,16 +572,11 @@ hash_timestamp_to_quanta(QField, QSize, QUnit, QIndex, MaxSubQueries, Where1) ->
+    %% use the maximum value that has not been incremented, we still use
+    %% the end_inclusive flag because the end key is not used to hash
+    make_wheres(Where2, QField, Min2, Max1, Boundaries).
 
```
There should be unit tests for queries with `ORDER BY` and `LIMIT` in here - we are not testing the `ORDER BY` compiler...

hmmr replied:
Added more tests in 2abb433.

gordonguthrie commented on src/riak_kv_qry_queue.erl:
```diff
@@ -67,18 +67,18 @@
 %%%===================================================================
 
--spec put_on_queue(pid(), [qry()], ?DDL{}) ->
+-spec put_on_queue(pid(), [qry()], ?DDL{}, riak_kv_qry_buffers:qbuf_ref() | undefined) ->
```
not that happy about using `undefined` to mean `no local query buffer`

hmmr replied:
But.. why?

gordonguthrie replied:
When you see `undefined` during debugging you don't know if it is an intended or unintended value...

gordonguthrie commented on src/riak_kv_qry_worker.erl:
```diff
@@ -134,9 +152,9 @@ code_change(_OldVsn, State, _Extra) ->
-run_sub_qs_fn([]) ->
-    ok;
-run_sub_qs_fn([{{qry, ?SQL_SELECT{cover_context = undefined} = Q1}, {qid, QId}} | T]) ->
+prepare_fsm_options([], Acc) ->
```
Sub-queries are always of one type so this endless partial recalculation is a bit redundant, no?

hmmr replied:
Refactored in 178cbdf.

Commit:
qry_buffers: thread ldb & oom errors; make qbufs non-optional for ORDER BY

Commit:
qry_buffers: get/set max_query_data_size

Andalso, correct the Reason atom for the associated error

Commit:
qry_buffers: create root dir for qbufs at init

Commit:
qry_buffers: fix a spec

Commit:
qry_worker: fix up use of #state.orig_qry

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

gordonguthrie commented on src/riak_kv_select.erl:
```diff
@@ -117,19 +134,36 @@ downgrade_select([v2,v1=To|Tail], Select1) ->
+    downgrade_select([To|Tail], Select2);
+downgrade_select([v3,v2=To|Tail], Select1) ->
+    Select2 = #riak_select_v2{
+        'SELECT'      = Select1#riak_select_v3.'SELECT',
```
**Note** check if the record formats are moved around (they are aren't they) and is the upgrade/downgrade OK

hmmr replied:
Aaand it did cause trouble, was bound to. When a 1.4 node receives an assignment to spin up an index_fsm, it eventually trips on bad `?SQL_SELECT` field index (around here https://github.com/basho/riak_kv/blob/riak_ts-develop/src/riak_kv_qry_coverage_plan.erl#L80), leading to a cryptic crash with these.. words:

```
{function_clause,
 [{riak_ql_ddl, make_key,
   [{key_v1, [{param_v1, [<<"a">>]}, {hash_fn_v1, riak_ql_quanta, quantum, [{param_v1, [<<"b">>]}, 10, s], timestamp}]},
    true,
    [{<<"a">>, 1}, {<<"b">>, 1000}]],
   [{file, "src/riak_ql_ddl.erl"}, {line, 168}]},
  {riak_kv_qry_coverage_plan, make_key, 3, [{file, "src/riak_kv_qry_coverage_plan.erl"}, {line, 129}]},
  {riak_kv_qry_coverage_plan, do_plan, 7, [{file, "src/riak_kv_qry_coverage_plan.erl"}, {line, 83}]},
  {riak_kv_index_fsm, create_plan, 8, [{file, "src/riak_kv_index_fsm.erl"}, {line, 161}]},
  {riak_core_coverage_fsm, initialize, 2, [{file, "src/riak_core_coverage_fsm.erl"}, {line, 272}]},
  {gen_fsm, handle_msg, 7, [{file, "gen_fsm.erl"}, {line, 505}]}, {proc_lib, init_p_do_apply, 3, [{file, "proc_lib.erl"}, {line, 239}]}]}
```

A fix is to append, rather than insert, new fields to `#riak_select_v3` (a7eb41a),

gordonguthrie replied:
Sorry, this is not acceptable - the system cannot work like this where the position of a key in a record at compile time changes the behaviour...

gordonguthrie commented on src/riak_kv_qry_buffers.erl:
```diff
@@ -0,0 +1,865 @@
+%% store OrderingDirection in ?SQL_PARAM where it belongs. Until then,
+%% we keep the ordering qualifiers in a makeshift #qbuf{} field outside DDL.
+make_lk_from_orderby(OrderBy) ->
+    #key_v1{ast = [?SQL_PARAM{name = [Field]} || {Field, _Dir, _NullsGroup} <- OrderBy]}.
```
`Dir` is a confusing abbreviation for `Direction` as I read it as `Directory`

hmmr replied:
Renamed (b8ed326e).

gordonguthrie commented on src/riak_kv_qry.erl:
```diff
@@ -281,22 +281,60 @@ count_to_position(Col, [_ | Rest], Pos) ->
     end.
 
+
+-spec maybe_create_query_buffer(?SQL_SELECT{}, pos_integer(), ?DDL{}, proplists:proplist()) ->
```
the word `maybe` in a function name usually indicates a side-effect which may or may not happen - is it right here? is `undefined` really an appropriate return value from this function?

hmmr replied:
Yes, `maybe_create_query_buffer` has the intended effect for eligible queries (and is a noop for others).

Re `undefined` for the `QBufRef` it returns, it's a way to avoid two extra indent levels. An alternative would be to have a case statement differentiating on `?SQL_SELECT{'ORDER BY' = DefinedOrNot}`, where `create_query_buffer` would do its work unconditionally. And, the `FullCycle` and downstream functions have been taught to handle `undefined` for `QBufRef`.

gordonguthrie commented on src/riak_kv_qry_buffers_ldb.erl:
```diff
@@ -0,0 +1,173 @@
+    FetchLimitFn =
+        fun(_KV, {Off, Lim, Pos, Acc}) when Pos < Off ->
+                {Off, Lim, Pos + 1, Acc};
+           %% Fetching K (let alone V) while "seeking" to our Offset
```
I am not sure this is a correct description of the seek function in leveldb - we need to raise this with @matthewvon and seek his opinion before embarking on this

hmmr replied:
Ideally, we would like to keep `{Pos, Iter}` pairs in `#qbuf` records, to seek to OFFSET positions from the nearest cached pair. As I point out in the comment to this piece of code, there is no iterator support in leveldb (at any rate, not with the expected semantics). I have confirmed this sad fact with @erikleitch.

For @matthewvon (or rather, for @erikleitch) to consider implementing iterators that can be exposed via external leveldb API, I would like to have (1) `fold` to accept an iterator as an argument, to have it start folding from that position, and (2) a `seek(Iter, Distance)` function that would return an iterator pointing to a position some `Distance` (positive or negative) relative to `Iter`, and maybe (3) the `fold` function to also return the iterator where it has finished.

Of course it is painful to watch `eleveldb:fold` fetch objects, only to be dropped as it iterates from the beginning, every damn time it is called. I really don't see how it can be instructed not to do so -- but neither do I want query buffers to be delayed by our wishful thinking.

hmmr replied:
Poring over https://github.com/basho/eleveldb/blob/riak_ts-develop/src/eleveldb.erl#L376L377, I start to wonder if I could cache `{Pos, Key}` pairs, where each `Key` would be extracted with `lists:last(RecordsForClient)`.. Still, it quacks like a kludge.

matthewvon replied:
@hmmr  API for manual manipulation of iterators within Erlang exists.  See riak_core/hashtree.erl for examples.  That code is already doing much of what you are requesting.

andytill commented on include/riak_kv_ts.hrl:
```diff
@@ -54,8 +54,8 @@
-          'ORDER BY'    = []    :: [riak_kv_qry_compiler:sorter()],
-          'LIMIT'       = []    :: [riak_kv_qry_compiler:limit()],
+          'ORDER BY'            :: undefined | [riak_kv_qry_compiler:sorter()],
+          'LIMIT'               :: undefined | riak_kv_qry_compiler:limit(),
```
We shouldn't change old records, because they still don't change on the old versioned nodes that create them.

andytill commented on include/riak_kv_ts.hrl:
```diff
@@ -71,8 +71,8 @@
-          'ORDER BY'    = []    :: [riak_kv_qry_compiler:sorter()],
-          'LIMIT'       = []    :: [riak_kv_qry_compiler:limit()],
+          'ORDER BY'            :: undefined | [riak_kv_qry_compiler:sorter()],
+          'LIMIT'               :: undefined | riak_kv_qry_compiler:limit(),
```
We shouldn't change old records, because they still don't change on the old versioned nodes that create them.

hmmr replied:
The composition of the record did not change, only the default values did. Because 1.4 nodes don't inspect `'ORDER BY'` and `'LIMIT'`, I don't expect things will break because of this change. An upgrade/downgrade test is in the plans.

gordonguthrie reviewed (commented):
Lots of not very major comments about style and readabillity - couple of major things not covered.

First up (as raised seperately) the code can be made to blow up by querying a comedy amount of quanta.

Secondly we don't have a complete set of error conditions - which we will need for documentation, training and review of the riak_test suites.

Commit:
qry_compiler: cap number of subqueries to a configurable value

Commit:
qry_compiler: diagnose ORDER BY with aggregate functions

Commit:
dialyzer fix after bfb275b5

Commit:
qry_buffers: delete unused internal API functions

because they are unused.

Commit:
qry_buffers: expose parameters via riak_kv.schema

these ones:

  %% max number of quanta (= subqueries) a query can span
  {timeseries_query_max_quanta_span, 5000},
  %% max data size a query can return to the client
  {timeseries_query_max_returned_data_size, 1000000},
  %% query buffers configuration:
  %% where query buffers will be created
  {timeseries_query_buffers_root_path, ""},
  %% query buffers soft and hard watermark
  {timeseries_query_buffers_soft_watermark, 1073741824},
  {timeseries_query_buffers_hard_watermark, 4294967296},
  %% expiration times
  {timeseries_query_buffers_expire_ms, 5000},
  {timeseries_query_buffers_incomplete_release_ms, 9000},

Commit:
riak_kv.schema: ensure query buffers root_path is absolute, normalized

Commit:
qry_buffers: use own rm -rf

Commit:
qry_buffers: do a gen_server:cast instead of ! in schedule_tick

matthewvon commented:
What is scenario that requires "rm -rf"?  Was there a crash?  Did you try API DestroyDB first to validate Db is not in use?

Commit:
qry_buffers: drop unused arg in schedule_tick

Commit:
ts_util: it's ok for rm_rf to see enoent on the Dir being deleted

Commit:
qry_buffers: let prepare_qbuf_dir work asynchronously, not delay init

hmmr commented:
@matthewvon After a crash, we certainly don't need stale temporary tables, because the queries for which they were created are gone. And, the application which does the clean-up at startup, is the only user of these tables, so we are certain no-one else will need them. Plain and brutal rm -rf is just appropriate here.

Commit:
qry_buffers: don't sort ?SQL_SELECT.'WHERE' for make_qref

because its contents is canonicalised.

Commit:
qry_buffers: replace if statements with case

Commit:
qry_buffers: avoid recompiling queries when building temp table DDL

pass CompiledSelect and CompiledOrderBy from riak_kv_qry:do_select.

Commit:
qry_buffers: define a macro for timer tick

Commit:
qry_buffers: don't pass Pid of self() for notification purposes

pass a function instead.

But.. why?

Commit:
qry_buffers: rename some variables for clarity

Commit:
qry_buffers: rename some functions for clarity

Commit:
qry_buffers_ldb: simplify call to eleveldb:fold in fetch_rows

Commit:
qry_compiler: add more tests for ORDER BY

with some formatting touchups

Commit:
riak_kv.schema: fixup after 80ace362

hmmr commented:
After a crash, we certainly don't need stale temporary tables, because
the queries for which they were created are gone. And, the application
which does the clean-up at startup, is the only user of these tables,
so we are certain no-one else will need them. Plain and brutal rm -rf
is just appropriate here.

Commit:
qry_worker: refactor prepare_fsm_options

Commit:
riak_kv.schema: don't set upper limit for query_max_returned_data_size

For the record, in ts_cluster_coverage, SELECTed size already reaches 7M.

Commit:
qry_buffers: double-ensure make_qbuf_id includes unique timestamps

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

Commit:
riak_kv_ts.hrl: only append new fields to next-version records

not insert before last.

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

# Conflicts:
#	src/riak_kv_ts_error_msgs.hrl

Commit:
qry_compiler: fix order_by_with_pass_1_test to match changes in ql

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

# Conflicts:
#	src/riak_kv_ts_svc.erl
#	src/riak_kv_ts_util.erl

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

# Conflicts:
#	src/riak_kv_pb_ts.erl
#	src/riak_kv_qry_compiler.erl
#	src/riak_kv_ttb_ts.erl

Commit:
ts_util: don't crash on non-existing tables in is_table_supported

Commit:
qry_compiler: fixups in tests after 4cca3efb

Commit:
Merge branch 'riak_ts-develop' into feature-az-query_temp_tables_and_snapshots

Commit:
compile_tab: when saving ddl_v1, don't forget to supply a Pid

this prevents riak from crashing at init time.

Commit:
Revert "ts_util: don't crash on non-existing tables in is_table_supported"

This reverts commit ce00a67712e9010676b2c08693dede664c0e1b2a.

Commit:
make 'ORDER BY', 'LIMIT', 'OFFSET' fields lists, for futureproofing

gordonguthrie commented:
+1 4ee8edd

hmmr commented:
@borshop retry

hmmr commented:
@borshop: retry

hmmr commented:
@borshop merge

I have manually initiated a rebuild, and it's green: http://buildbot.bos1/builders/test-riak_kv/builds/4250.

hmmr commented:
@borshop: merge

hmmr commented:
@borshop: retry 4ee8edd

Commit:
remove a duplicate 'assignment'

Commit:
select: avoid addressing record fields by indices when up/downgrading

andytill commented:
+1 0658863

hmmr commented:
@borshop merge

hmmr commented:
@borshop merge

hmmr commented:
@borshop merge

hmmr commented:
@borshop: merge
